### PR TITLE
non-blocking iterator over entire index

### PR DESCRIPTION
**Pull request by rzezeski**
This is needed for async folding which is needed for the new repair functionality which deals with only primary vnodes.  You don't want repair to block a vnode from handling other requests.

### CODE REVIEW

File: src/mi_server.erl [hunk 1]
```diff
@@ -370,6 +378,42 @@ handle_call(is_empty, _From, State) ->
...
+
+    F2 = fun(Segment, Acc) ->
+                 mi_locks:claim(mi_segment:filename(Segment), Acc)
+         end,
+    NewLocks1 = lists:foldl(F2, NewLocks, Segments),
+
+    %% build ordered iterator over all buffers + segments
+    %%
+    %% TODO: maybe don't bother with order here since this is for
+    %% repair?
```
Comment (slfritchie) [inline src/mi_server.erl]: Unless the order is extremely expensive computation-wise or I/O-wise, I suggest removing the TODO comment: other functions will probably use this feature in later releases.

File: src/mi_server.erl [hunk 2]
```diff
@@ -370,6 +378,42 @@ handle_call(is_empty, _From, State) ->
...
+    %% being iterated
+    F1 = fun(Buffer, Acc) ->
+                 mi_locks:claim(mi_buffer:filename(Buffer), Acc)
+         end,
+    NewLocks = lists:foldl(F1, Locks, Buffers),
+
+    F2 = fun(Segment, Acc) ->
+                 mi_locks:claim(mi_segment:filename(Segment), Acc)
+         end,
+    NewLocks1 = lists:foldl(F2, NewLocks, Segments),
```
Comment (slfritchie) [inline src/mi_server.erl]: This acquisition of buffer & segment locks happens multiple times in similar/identical code, seems like a DRY refactoring opportunity?
- lookup
- range
- iterator (the new addition)

File: src/mi_server.erl (hunk 1 above)
Comment (rzezeski) [inline src/mi_server.erl]: Removed.

File: src/mi_server.erl (hunk 2 above)
Comment (rzezeski) [inline src/mi_server.erl]: Been meaning to do this for a while, done.

**Comment by slfritchie** [comment]
How worried should we be that the async fold worker/servers's protocol for sending results to the client is strictly unidirectional and doesn't have any backpressure mechanism?

**Comment by rzezeski** [comment]
Scott, I think this is a concern since it's feeding the handoff
sender which has back-pressure and must send over the network.
This same pipeline is used during queries but at least there the
pipeline probably flows better (I'm making an assumption here as
this can also traverse network).  If the index is large enough
and the disproportion between the handoff xfer and local iterator
is great enough then the mailbox/process heap on the vnode worker
will grow and potentially cause a crash

I could modify `iterate2` to only send N result chunks before
waiting for a CONTINUE and make a new `result_iterator` that
sends a CONTINUE after each chunk it reads.  The reason for N
chunks is so that some data is preemptively loaded so each call
to iterator doesn't have to wait for msg req/response cycle.
Does that make sense?  What do you think?

**Comment by slfritchie** [comment]
Ryan, I think that your proposal is a good one.  It isn't necessary for `iterate2` and the result iterator thing to be in lock-step with the backpressure from the handoff receiver & sender.  But I think it is necessary for the MI iterator to eventually stop sending to the handoff sender if the handoff sender stops sending.

**Comment by rzezeski** [comment]
Yep, I agree that they don't need to be lock-step.  In fact, I'd consider back-pressure in MI to be orthogonal and good for _all_ cases of iteration including queries.  However, for now I'll only update the `iterator` API and leave a TODO in the code about adding back-pressure to all MI APIs.  That way I can have more time to adequately test for any performance regressions whereas with repair/handoff I prefer robustness (queries already sorta have that since their scope is limited).

**Pull request merged**